File: app/src/main/java/com/apps/adrcotfas/goodtime/BL/TimerService.java
/*
 * Copyright 2016-2020 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.apps.adrcotfas.goodtime.BL;

import android.annotation.TargetApi;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.media.AudioManager;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.Handler;
import android.os.PowerManager;
import android.util.Log;

import com.apps.adrcotfas.goodtime.Database.AppDatabase;
import com.apps.adrcotfas.goodtime.Main.TimerActivity;
import com.apps.adrcotfas.goodtime.Session;
import com.apps.adrcotfas.goodtime.Settings.PreferenceHelper;
import com.apps.adrcotfas.goodtime.Util.Constants;

import java.util.concurrent.TimeUnit;

import androidx.annotation.RequiresApi;
import androidx.lifecycle.LifecycleService;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import static android.media.AudioManager.RINGER_MODE_SILENT;
import static com.apps.adrcotfas.goodtime.BL.NotificationHelper.GOODTIME_NOTIFICATION_ID;
import static com.apps.adrcotfas.goodtime.Util.Constants.SESSION_TYPE;
import static com.apps.adrcotfas.goodtime.Util.StringUtils.formatTime;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

/**
 * Class representing the foreground service which triggers the countdown timer and handles events.
 */
@AndroidEntryPoint
public class TimerService extends LifecycleService {

    private static final String TAG = TimerService.class.getSimpleName();

    private NotificationHelper notificationHelper;

    @Inject
    RingtoneAndVibrationPlayer ringtoneAndVibrationPlayer;
    
    @Inject PreferenceHelper preferenceHelper;
    
    @Inject CurrentSessionManager currentSessionManager;

    private int previousRingerMode;
    private boolean previousWifiMode;

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "onCreate " + this.hashCode());
        notificationHelper = new NotificationHelper(getApplicationContext());
        EventBus.getDefault().register(this);
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy " + this.hashCode());
        EventBus.getDefault().unregister(this);
        super.onDestroy();
    }

    @Override
    public synchronized int onStartCommand(final Intent intent, int flags, int startId) {
        super.onStartCommand(intent, flags, startId);

        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }

        int result = START_STICKY;

        if (intent == null) {
            return result;
        }

        Log.d(TAG, "onStartCommand " + this.hashCode() + " " + intent.getAction());
        switch (intent.getAction()) {
            case Constants.ACTION.STOP:
                onStopEvent();
                break;
            case Constants.ACTION.TOGGLE:
                onToggleEvent();
                result = START_NOT_STICKY;
                break;
            case Constants.ACTION.START:
                SessionType sessionType = SessionType.valueOf(intent.getStringExtra(SESSION_TYPE));
                onStartEvent(sessionType);
                break;
            case Constants.ACTION.ADD_SECONDS:
                onAdd60Seconds();
                break;
            case Constants.ACTION.SKIP:
                onSkipEvent();
                break;
            default:
                break;
        }
        return result;
    }

    /**
     * Called when an event is posted to the EventBus
     * @param o holds the type of the Event
     */
    @Subscribe
    public void onEvent(Object o) {
        if (o instanceof Constants.OneMinuteLeft) {
            onOneMinuteLeft();
        } else if (o instanceof Constants.FinishWorkEvent) {
            Log.d(TAG, "onEvent " + o.getClass().getSimpleName());
            onFinishEvent(SessionType.WORK);
        } else if (o instanceof Constants.FinishBreakEvent) {
            Log.d(TAG, "onEvent " + o.getClass().getSimpleName());
            onFinishEvent(SessionType.BREAK);
        } else if (o instanceof  Constants.FinishLongBreakEvent) {
            onFinishEvent(SessionType.LONG_BREAK);
        } else if (o instanceof Constants.UpdateTimerProgressEvent) {
            updateNotificationProgress();
        } else if (o instanceof Constants.ClearNotificationEvent) {
            Log.d(TAG, "onEvent " + o.getClass().getSimpleName());
            notificationHelper.clearNotification();
            ringtoneAndVibrationPlayer.stop();
        }
    }

    private void onStartEvent(SessionType sessionType) {

        EventBus.getDefault().post(new Constants.StartSessionEvent());
        if (sessionType != SessionType.WORK && preferenceHelper.isLongBreakEnabled()
                && preferenceHelper.itsTimeForLongBreak()) {
            sessionType = SessionType.LONG_BREAK;
        }

        Log.d(TAG, "onStartEvent: " + sessionType.toString());

        currentSessionManager.startTimer(sessionType);

        if (sessionType == SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(false);
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(false);
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(false);
            }
        }

        if (!preferenceHelper.isAutoStartWork() && !preferenceHelper.isAutoStartBreak()) {
            ringtoneAndVibrationPlayer.stop();
        }

        notificationHelper.clearNotification();
        startForeground(GOODTIME_NOTIFICATION_ID, notificationHelper.getInProgressBuilder(
                currentSessionManager.getCurrentSession()).build());
    }

    private void onToggleEvent() {
        currentSessionManager.toggleTimer();
        startForeground(GOODTIME_NOTIFICATION_ID, notificationHelper.getInProgressBuilder(
                currentSessionManager.getCurrentSession()).build());
    }

    private void onStopEvent() {
        Log.d(TAG, "onStopEvent");

        if (preferenceHelper.isWiFiDisabled()) {
            toggleWifi(true);
        }
        if (preferenceHelper.isSoundAndVibrationDisabled()) {
            toggleSound(true);
        }
        if (preferenceHelper.isDndModeActive()) {
            toggleDndMode(true);
        }

        SessionType sessionType = currentSessionManager.getCurrentSession().getSessionType().getValue();
        Log.d(TAG, "onStopEvent, sessionType: " + sessionType);
        if (sessionType == SessionType.LONG_BREAK) {
            preferenceHelper.resetCurrentStreak();
        }

        stopForeground(true);
        stopSelf();

        finalizeSession(sessionType, currentSessionManager.getElapsedMinutesAtStop());
    }

    private void onOneMinuteLeft() {
        acquireScreenLock();
        bringActivityToFront();
        ringtoneAndVibrationPlayer.play(SessionType.WORK, false);
    }

    private void onFinishEvent(SessionType sessionType) {
        Log.d(TAG, TimerService.this.hashCode() + " onFinishEvent " + sessionType.toString());

        acquireScreenLock();
        bringActivityToFront();

        if (sessionType == SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(true);
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(true);
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(true);
            }
        }

        ringtoneAndVibrationPlayer.play(sessionType, preferenceHelper.isRingtoneInsistent());
        stopForeground(true);

        updateLongBreakStreak(sessionType);

        // store what was done to the database
        finalizeSession(sessionType, currentSessionManager.getElapsedMinutesAtFinished());

        if (preferenceHelper.isAutoStartBreak() && sessionType == SessionType.WORK) {
            onStartEvent(SessionType.BREAK);
        } else if (preferenceHelper.isAutoStartWork() && sessionType != SessionType.WORK) {
            onStartEvent(SessionType.WORK);
        } else {
            notificationHelper.notifyFinished(sessionType);
        }
    }

    private void onAdd60Seconds() {
        Log.d(TAG, TimerService.this.hashCode() + " onAdd60Seconds ");
        preferenceHelper.increment60SecondsCounter();
        if (currentSessionManager.getCurrentSession().getTimerState().getValue() == TimerState.INACTIVE) {
            startForeground(GOODTIME_NOTIFICATION_ID, notificationHelper.getInProgressBuilder(
                    currentSessionManager.getCurrentSession()).build());
        }
        currentSessionManager.add60Seconds();
    }

    private void onSkipEvent() {
        final SessionType sessionType = currentSessionManager.getCurrentSession().getSessionType().getValue();
        Log.d(TAG, TimerService.this.hashCode() + " onSkipEvent " + sessionType.toString());

        if (sessionType == SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(true);
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(true);
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(true);
            }
        }

        currentSessionManager.stopTimer();
        stopForeground(true);
        updateLongBreakStreak(sessionType);

        finalizeSession(sessionType, currentSessionManager.getElapsedMinutesAtStop());

        onStartEvent(sessionType == SessionType.WORK ? SessionType.BREAK : SessionType.WORK);
    }

    private void updateLongBreakStreak(SessionType sessionType) {
        if (preferenceHelper.isLongBreakEnabled()) {
            if (sessionType == SessionType.LONG_BREAK) {
                preferenceHelper.resetCurrentStreak();
            } else if (sessionType == SessionType.WORK){
                preferenceHelper.incrementCurrentStreak();
            }
            Log.d(TAG, "preferenceHelper.getCurrentStreak: " + preferenceHelper.getCurrentStreak());
            Log.d(TAG, "preferenceHelper.lastWorkFinishedAt: " + preferenceHelper.lastWorkFinishedAt());
        }
    }

    private void acquireScreenLock() {
        PowerManager powerManager = (PowerManager) getSystemService(POWER_SERVICE);
        PowerManager.WakeLock wakeLock = powerManager.newWakeLock(PowerManager.FULL_WAKE_LOCK | PowerManager.ACQUIRE_CAUSES_WAKEUP,
                AlarmReceiver.class.getName());
        wakeLock.acquire(5000);
    }

    private void updateNotificationProgress() {
        notificationHelper.updateNotificationProgress(
                currentSessionManager.getCurrentSession());
    }

    private void toggleSound(boolean restore) {
        if (isNotificationPolicyAccessGranted()) {
            toggleSoundInternal(restore);
        } else {
            // should not happen
            Log.w(TAG, "Trying to toggle sound but permission was not granted.");
        }
    }

    private void toggleSoundInternal(boolean restore) {
        Thread t = new Thread(() -> {
            final AudioManager aManager = (AudioManager) getSystemService(AUDIO_SERVICE);
            if (restore) {
                if (previousRingerMode == RINGER_MODE_SILENT) {
                    return;
                }
                aManager.setRingerMode(previousRingerMode);
            } else {
                previousRingerMode = aManager.getRingerMode();
                aManager.setRingerMode(RINGER_MODE_SILENT);
            }
        });
        t.start();
    }

    private void toggleDndMode(boolean restore) {
        if (isNotificationPolicyAccessGranted()) {
            togglePriorityMode(restore);
        } else {
            // should not happen
            Log.w(TAG, "Trying to toggle DnD mode but permission was not granted.");
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private void togglePriorityMode(boolean restore) {
        Thread t = new Thread(() -> {
            NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            if (manager != null) {
                if (restore) {
                    manager.setInterruptionFilter(NotificationManager.INTERRUPTION_FILTER_ALL);
                } else {
                    manager.setInterruptionFilter(NotificationManager.INTERRUPTION_FILTER_PRIORITY);
                }
            }
        });
        t.start();
    }

    private void toggleWifi(boolean restore) {
        Runnable r = () -> {
            WifiManager wifiManager = (WifiManager) this.getSystemService(WIFI_SERVICE);
            if (restore) {
                wifiManager.setWifiEnabled(previousWifiMode);
            } else {
                previousWifiMode = wifiManager.isWifiEnabled();
                wifiManager.setWifiEnabled(false);
            }
        };
        Thread t = new Thread(r);
        t.start();
    }

    //TODO: clean-up this mess
    private void finalizeSession(SessionType sessionType, int minutes) {
        currentSessionManager.stopTimer();
        preferenceHelper.resetAdd60SecondsCounter();
        currentSessionManager.getCurrentSession().setDuration(
                TimeUnit.MINUTES.toMillis(preferenceHelper.getSessionDuration(SessionType.WORK)));

        if (sessionType != SessionType.WORK) {
            return;
        }

        final String label = currentSessionManager.getCurrentSession().getLabel().getValue();
        final long endTime = System.currentTimeMillis();

        Handler handler = new Handler();

        Runnable r = () -> {
            Log.d(TAG, "finalizeSession / elapsed minutes: " + minutes);
            if (minutes > 0) {
                try {
                    Session session = new Session(
                            0,
                            endTime,
                            minutes,
                            label);
                    AppDatabase.getDatabase(getApplicationContext()).sessionModel().addSession(session);
                    Log.d(TAG, "finalizeSession, saving session finished at" + formatTime(endTime));
                } catch (Exception e) {
                    // the label was deleted in the meantime so set it to null and save the unlabeled session
                    handler.post(() -> currentSessionManager.getCurrentSession().setLabel(null));
                    Session session = new Session(
                            0,
                            endTime,
                            minutes,
                            null);
                    AppDatabase.getDatabase(getApplicationContext()).sessionModel().addSession(session);
                }
            }
        };
        Thread t = new Thread(r);
        Log.d(TAG, "finalizeSession, start thread");
        t.start();
    }

    private void bringActivityToFront() {
        Intent activityIntent = new Intent(this, TimerActivity.class);
        activityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        activityIntent.putExtra("42", 42);
        getApplication().startActivity(activityIntent);
    }

    @TargetApi(Build.VERSION_CODES.M)
    private boolean isNotificationPolicyAccessGranted() {
        NotificationManager notificationManager = (NotificationManager)
                getSystemService(Context.NOTIFICATION_SERVICE);
        return notificationManager.isNotificationPolicyAccessGranted();
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/bl/TimerService.kt
/*
 * Copyright 2016-2021 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.bl

import com.apps.adrcotfas.goodtime.database.AppDatabase.Companion.getDatabase
import dagger.hilt.android.AndroidEntryPoint
import androidx.lifecycle.LifecycleService
import javax.inject.Inject
import com.apps.adrcotfas.goodtime.settings.PreferenceHelper
import org.greenrobot.eventbus.EventBus
import kotlin.jvm.Synchronized
import android.content.Intent
import org.greenrobot.eventbus.Subscribe
import com.apps.adrcotfas.goodtime.util.Constants.OneMinuteLeft
import com.apps.adrcotfas.goodtime.util.Constants.FinishWorkEvent
import com.apps.adrcotfas.goodtime.util.Constants.FinishBreakEvent
import com.apps.adrcotfas.goodtime.util.Constants.FinishLongBreakEvent
import com.apps.adrcotfas.goodtime.util.Constants.UpdateTimerProgressEvent
import com.apps.adrcotfas.goodtime.util.Constants.ClearNotificationEvent
import com.apps.adrcotfas.goodtime.util.Constants.StartSessionEvent
import android.os.PowerManager
import android.media.AudioManager
import androidx.annotation.RequiresApi
import android.os.Build
import android.app.NotificationManager
import android.net.wifi.WifiManager
import com.apps.adrcotfas.goodtime.main.TimerActivity
import android.annotation.TargetApi
import android.util.Log
import androidx.lifecycle.lifecycleScope
import com.apps.adrcotfas.goodtime.database.Session
import com.apps.adrcotfas.goodtime.util.Constants
import com.apps.adrcotfas.goodtime.util.toFormattedTime
import com.apps.adrcotfas.goodtime.util.toLocalTime
import kotlinx.coroutines.*
import java.lang.Exception
import java.lang.Runnable
import java.util.concurrent.TimeUnit

/**
 * Class representing the foreground service which triggers the countdown timer and handles events.
 */
@AndroidEntryPoint
class TimerService : LifecycleService() {

    @Inject
    lateinit var notificationHelper: NotificationHelper

    @Inject
    lateinit var ringtoneAndVibrationPlayer: RingtoneAndVibrationPlayer

    @Inject
    lateinit var preferenceHelper: PreferenceHelper

    @Inject
    lateinit var currentSessionManager: CurrentSessionManager

    private var previousRingerMode = 0
    private var previousWifiMode = false

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, "onCreate " + this.hashCode())
        notificationHelper = NotificationHelper(applicationContext)
        EventBus.getDefault().register(this)
    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy " + this.hashCode())
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }

    @Synchronized
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
        var result = START_STICKY
        if (intent == null) {
            return result
        }
        Log.d(TAG, "onStartCommand " + this.hashCode() + " " + intent.action)
        when (intent.action) {
            Constants.ACTION.STOP -> onStopEvent()
            Constants.ACTION.TOGGLE -> {
                onToggleEvent()
                result = START_NOT_STICKY
            }
            Constants.ACTION.START -> {
                val sessionType =
                    SessionType.valueOf(intent.getStringExtra(Constants.SESSION_TYPE)!!)
                onStartEvent(sessionType)
            }
            Constants.ACTION.ADD_SECONDS -> onAdd60Seconds()
            Constants.ACTION.SKIP -> onSkipEvent()
            else -> {
            }
        }
        return result
    }

    /**
     * Called when an event is posted to the EventBus
     * @param o holds the type of the Event
     */
    @Subscribe
    fun onEvent(o: Any) {
        when (o) {
            is OneMinuteLeft -> {
                onOneMinuteLeft()
            }
            is FinishWorkEvent -> {
                Log.d(TAG, "onEvent " + o.javaClass.simpleName)
                onFinishEvent(SessionType.WORK)
            }
            is FinishBreakEvent -> {
                Log.d(TAG, "onEvent " + o.javaClass.simpleName)
                onFinishEvent(SessionType.BREAK)
            }
            is FinishLongBreakEvent -> {
                onFinishEvent(SessionType.LONG_BREAK)
            }
            is UpdateTimerProgressEvent -> {
                updateNotificationProgress()
            }
            is ClearNotificationEvent -> {
                Log.d(TAG, "onEvent " + o.javaClass.simpleName)
                notificationHelper.clearNotification()
                ringtoneAndVibrationPlayer.stop()
            }
        }
    }

    private fun onStartEvent(sessionType: SessionType) {
        var sessionTypeTmp = sessionType
        EventBus.getDefault().post(StartSessionEvent())
        if (sessionTypeTmp !== SessionType.WORK && preferenceHelper.isLongBreakEnabled()
            && preferenceHelper.itsTimeForLongBreak()
        ) {
            sessionTypeTmp = SessionType.LONG_BREAK
        }
        Log.d(TAG, "onStartEvent: $sessionTypeTmp")
        currentSessionManager.startTimer(sessionTypeTmp)
        if (sessionTypeTmp === SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(false)
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(false)
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(false)
            }
        }
        if (!preferenceHelper.isAutoStartWork() && !preferenceHelper.isAutoStartBreak()) {
            ringtoneAndVibrationPlayer.stop()
        }
        notificationHelper.clearNotification()
        startForeground(
            NotificationHelper.GOODTIME_NOTIFICATION_ID, notificationHelper.getInProgressBuilder(
                currentSessionManager.currentSession
            ).build()
        )
    }

    private fun onToggleEvent() {
        currentSessionManager.toggleTimer()
        startForeground(
            NotificationHelper.GOODTIME_NOTIFICATION_ID, notificationHelper.getInProgressBuilder(
                currentSessionManager.currentSession
            ).build()
        )
    }

    private fun onStopEvent() {
        Log.d(TAG, "onStopEvent")
        if (preferenceHelper.isWiFiDisabled()) {
            toggleWifi(true)
        }
        if (preferenceHelper.isSoundAndVibrationDisabled()) {
            toggleSound(true)
        }
        if (preferenceHelper.isDndModeActive()) {
            toggleDndMode(true)
        }
        val sessionType = currentSessionManager.currentSession.sessionType.value
        Log.d(TAG, "onStopEvent, sessionType: $sessionType")
        if (sessionType === SessionType.LONG_BREAK) {
            preferenceHelper.resetCurrentStreak()
        }
        stopForeground()
        stopSelf()
        finalizeSession(sessionType, currentSessionManager.elapsedMinutesAtStop)
    }

    private fun onOneMinuteLeft() {
        acquireScreenLock()
        bringActivityToFront()
        ringtoneAndVibrationPlayer.play(SessionType.WORK, false)
    }

    private fun onFinishEvent(sessionType: SessionType) {
        Log.d(
            TAG,
            this@TimerService.hashCode().toString() + " onFinishEvent " + sessionType.toString()
        )
        acquireScreenLock()
        bringActivityToFront()
        if (sessionType === SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(true)
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(true)
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(true)
            }
        }
        ringtoneAndVibrationPlayer.play(sessionType, preferenceHelper.isRingtoneInsistent())
        stopForeground()
        updateLongBreakStreak(sessionType)

        // store what was done to the database
        finalizeSession(sessionType, currentSessionManager.elapsedMinutesAtFinished)
        if (preferenceHelper.isAutoStartBreak() && sessionType === SessionType.WORK) {
            onStartEvent(SessionType.BREAK)
        } else if (preferenceHelper.isAutoStartWork() && sessionType !== SessionType.WORK) {
            onStartEvent(SessionType.WORK)
        } else {
            notificationHelper.notifyFinished(sessionType)
        }
    }

    private fun onAdd60Seconds() {
        Log.d(TAG, this@TimerService.hashCode().toString() + " onAdd60Seconds ")
        preferenceHelper.increment60SecondsCounter()
        if (currentSessionManager.currentSession.timerState.value == TimerState.INACTIVE) {
            startForeground(
                NotificationHelper.GOODTIME_NOTIFICATION_ID,
                notificationHelper.getInProgressBuilder(
                    currentSessionManager.currentSession
                ).build()
            )
        }
        currentSessionManager.add60Seconds()
    }

    private fun onSkipEvent() {
        val sessionType = currentSessionManager.currentSession.sessionType.value
        Log.d(
            TAG,
            this@TimerService.hashCode().toString() + " onSkipEvent " + sessionType.toString()
        )
        if (sessionType === SessionType.WORK) {
            if (preferenceHelper.isWiFiDisabled()) {
                toggleWifi(true)
            }
            if (preferenceHelper.isSoundAndVibrationDisabled()) {
                toggleSound(true)
            }
            if (preferenceHelper.isDndModeActive()) {
                toggleDndMode(true)
            }
        }
        currentSessionManager.stopTimer()
        stopForeground()
        updateLongBreakStreak(sessionType)
        finalizeSession(sessionType, currentSessionManager.elapsedMinutesAtStop)
        onStartEvent(if (sessionType === SessionType.WORK) SessionType.BREAK else SessionType.WORK)
    }

    private fun updateLongBreakStreak(sessionType: SessionType?) {
        if (preferenceHelper.isLongBreakEnabled()) {
            if (sessionType === SessionType.LONG_BREAK) {
                preferenceHelper.resetCurrentStreak()
            } else if (sessionType === SessionType.WORK) {
                preferenceHelper.incrementCurrentStreak()
            }
            Log.d(
                TAG,
                "preferenceHelper.getCurrentStreak: " + preferenceHelper.getCurrentStreak()
            )
            Log.d(
                TAG,
                "preferenceHelper.lastWorkFinishedAt: " + preferenceHelper.lastWorkFinishedAt()
            )
        }
    }

    private fun acquireScreenLock() {
        val powerManager = getSystemService(POWER_SERVICE) as PowerManager
        val wakeLock = powerManager.newWakeLock(
            PowerManager.FULL_WAKE_LOCK or PowerManager.ACQUIRE_CAUSES_WAKEUP,
            AlarmReceiver::class.java.name
        )
        wakeLock.acquire(5000)
    }

    private fun updateNotificationProgress() {
        notificationHelper.updateNotificationProgress(
            currentSessionManager.currentSession
        )
    }

    private fun toggleSound(restore: Boolean) {
        if (isNotificationPolicyAccessGranted) {
            toggleSoundInternal(restore)
        } else {
            // should not happen
            Log.w(TAG, "Trying to toggle sound but permission was not granted.")
        }
    }

    private fun toggleSoundInternal(restore: Boolean) {
        val t = Thread {
            val aManager = getSystemService(AUDIO_SERVICE) as AudioManager
            if (restore) {
                if (previousRingerMode == AudioManager.RINGER_MODE_SILENT) {
                    return@Thread
                }
                aManager.ringerMode = previousRingerMode
            } else {
                previousRingerMode = aManager.ringerMode
                aManager.ringerMode = AudioManager.RINGER_MODE_SILENT
            }
        }
        t.start()
    }

    private fun toggleDndMode(restore: Boolean) {
        if (isNotificationPolicyAccessGranted) {
            togglePriorityMode(restore)
        } else {
            // should not happen
            Log.w(TAG, "Trying to toggle DnD mode but permission was not granted.")
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private fun togglePriorityMode(restore: Boolean) {
        val t = Thread {
            val manager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            if (restore) {
                manager.setInterruptionFilter(NotificationManager.INTERRUPTION_FILTER_ALL)
            } else {
                manager.setInterruptionFilter(NotificationManager.INTERRUPTION_FILTER_PRIORITY)
            }
        }
        t.start()
    }

    private fun toggleWifi(restore: Boolean) {
        val r = Runnable {
            val wifiManager = this.getSystemService(WIFI_SERVICE) as WifiManager
            if (restore) {
                wifiManager.isWifiEnabled = previousWifiMode
            } else {
                previousWifiMode = wifiManager.isWifiEnabled
                wifiManager.isWifiEnabled = false
            }
        }
        val t = Thread(r)
        t.start()
    }

    private fun finalizeSession(sessionType: SessionType?, minutes: Int) {
        currentSessionManager.stopTimer()
        preferenceHelper.resetAdd60SecondsCounter()
        currentSessionManager.currentSession.setDuration(
            TimeUnit.MINUTES.toMillis(preferenceHelper.getSessionDuration(SessionType.WORK))
        )
        if (sessionType !== SessionType.WORK) {
            return
        }

        val labelVal = currentSessionManager.currentSession.label.value
        val labelValProper =
            if (labelVal == null || labelVal == "" || labelVal == "unlabeled") null else labelVal

        val endTime = System.currentTimeMillis()
        Log.d(TAG, "finalizeSession / elapsed minutes: $minutes")
        if (minutes > 0) {
            Log.d(
                TAG,
                "finalizeSession, saving session finished at" + endTime.toLocalTime()
                    .toFormattedTime()
            )
            val session = Session(0, endTime, minutes, labelValProper)
            lifecycleScope.launch {
                try {
                    getDatabase(applicationContext).sessionModel().addSession(session)
                } catch (e: Exception) {
                    // the label was deleted in the meantime so set it to null and save the unlabeled session
                    withContext(Dispatchers.Main) {
                        currentSessionManager.currentSession.setLabel("")
                    }
                    val newSession = session.apply { label = null }
                    getDatabase(applicationContext).sessionModel().addSession(newSession)
                }
            }
        }
    }

    private fun stopForeground() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O)
            stopForeground(STOP_FOREGROUND_REMOVE)
        else
            stopForeground(true)
    }

    private fun bringActivityToFront() {
        val activityIntent = Intent(this, TimerActivity::class.java)
        activityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        activityIntent.putExtra("42", 42)
        application.startActivity(activityIntent)
    }

    @get:TargetApi(Build.VERSION_CODES.M)
    private val isNotificationPolicyAccessGranted: Boolean
        get() {
            val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            return notificationManager.isNotificationPolicyAccessGranted
        }

    companion object {
        private val TAG = TimerService::class.java.simpleName
    }
}